File: v1-cpp-cli/src/SecureNotes.cpp
#include "SecureNotes.h"

void SecureNotes::handleSecureNotes(AppState& state, int& key) {
	int input;
	std::string title;
	std::cout << "\n Secure Notes" << std::endl;
	std::cout << "===Secure-Vault===" << std::endl;
	std::cout << "[1] Add Note" << std::endl;
	std::cout << "[2] Get Note" << std::endl;
	std::cout << "[3] Delete Note" << std::endl;
	std::cout << "[4] List Notes" << std::endl;
	std::cout << "[0] Return to menu\n" << std::endl;

	if (!(std::cin >> input)) {
		std::cin.clear();
		std::cin.ignore(10000, '\n');
		std::cout << "Invalid input.\n" << std::endl;
		return;
	}

	switch (input) {
	case 0:
		state = AppState::MAIN_MENU;
		return;
	case 1:
		addNote(key);
		return;
	case 2:
		std::cin.ignore(10000, '\n');
		std::cout << "Enter note title to retrieve: ";
		std::getline(std::cin, title);
		getNote(title, key);
		return;
	case 3:
		std::cin.ignore(10000, '\n');
		std::cout << "Enter note title to delete: ";
		std::getline(std::cin, title);
		deleteNote(title, key);
		return;
	case 4:
		listNotes(key);
		return;
	}
}

std::string SecureNotes::xorEncryptDecrypt(std::string& input, int key) {
	std::string output = input;
	for (int i = 0; i < input.size(); ++i) {
		output[i] = input[i] ^ key;
	}
	return output;
}

void SecureNotes::addNote(int& key) {
	std::string title, content;

	std::cin.ignore(10000, '\n');
	std::cout << "===Add Note===\n" << std::endl;
	std::cout << "Enter note title: ";
	std::getline(std::cin, title);
	std::cout << "Enter note content: ";
	std::getline(std::cin, content);

	NoteEntry newEntry = { title, content };
	notes.push_back(newEntry);

	std::ofstream outFile("notes.dat", std::ios::app);
	if (!outFile) {
		std::cout << "Error opening file.\n" << std::endl;
		return;
	}

	outFile << title << "|" << xorEncryptDecrypt(content, key) << "\n";
	outFile.close();
}

void SecureNotes::getNote(std::string& title, int& key) {
	std::string checkTitle, checkContent;
	std::ifstream inFile("notes.dat");

	if (!inFile) {
		std::cout << "Error opening file.\n" << std::endl;
		return;
	}

	while (inFile) {
		std::getline(inFile, checkTitle, '|');
		std::getline(inFile, checkContent);
		if (checkTitle == title) {
			std::string decryptedContent = xorEncryptDecrypt(checkContent, key);
			std::cout << "\nTitle: " << checkTitle << "\nContent: " << decryptedContent << "\n" << std::endl;
			return;
		}
	}

	std::cout << "\nNo entry found for that title.\n" << std::endl;
}

void SecureNotes::deleteNote(std::string& title, int& key) {
	std::string checkTitle, checkContent;
	std::vector<NoteEntry> tempNotes;
	bool found = false;

	std::ifstream inFile("notes.dat");
	if (!inFile) {
		std::cout << "Error opening file.\n" << std::endl;
		return;
	}

	while (inFile) {
		std::getline(inFile, checkTitle, '|');
		std::getline(inFile, checkContent);

		if (checkTitle != title) {
			std::string decrypted = xorEncryptDecrypt(checkContent, key);
			tempNotes.push_back({ checkTitle, decrypted });
		}
		else found = true;
	}
	inFile.close();
	
	std::ofstream outFile("notes.dat");
	if (!outFile) {
		std::cout << "Error opening file.\n" << std::endl;
		return;
	}

	for (const auto& note : tempNotes) {
		std::string content = note.content;
		outFile << note.title << "|" << xorEncryptDecrypt(content, key) << "\n";
	}

	if (found) std::cout << "\nNote deleted successfully.\n" << std::endl;
	else std::cout << "\nNo entry found for that title.\n" << std::endl;
}

void SecureNotes::listNotes(int& key) {
	std::string title, content;
	std::ifstream inFile("notes.dat");
	int counter = 0;

	if (!inFile) {
		std::cout << "Error opening file.\n" << std::endl;
		return;
	}
	std::cout << "\nYour Notes:\n" << std::endl;

	while (inFile) {
		std::getline(inFile, title, '|');
		std::getline(inFile, content);

		if (!title.empty()) {
			counter++;
			std::cout << "Title " << "[" << counter << "]: " << title << std::endl;
		}
	}
}

File: v2-csharp-cli/src/SecureNotes.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Text;
using static System.Net.Mime.MediaTypeNames;

namespace SecureVault
{
    public class SecureNotes
    {
        public struct NoteEntry
        {
            public string Title {  get; set; }
            public string Content { get; set; }
        }

        public void handleSecureNotes()
        {
            int input = -1;
            Console.WriteLine("\n Secure Notes");
            Console.WriteLine("===Secure-Vault===");
            Console.WriteLine("[1] Add Note");
            Console.WriteLine("[2] Get Note");
            Console.WriteLine("[3] List Note");
            Console.WriteLine("[4] Delete Note");
            Console.WriteLine("[0] Return to menu\n");

            try { input = int.Parse(Console.ReadLine() ); } 
            catch (Exception) { Console.WriteLine("Invalid input.\n");}

            switch (input)
            {
                case 0:
                    App.state = AppState.MAIN_MENU;
                    break;
                case 1:
                    addNote();
                    break;
                case 2:
                    getNote();
                    break;
                case 3:
                    listNotes();
                    break;
                case 4:
                    deleteNote();
                    break;
            }

        }
        public void addNote()
        {
            string title, content, encrypted_content;

            Console.WriteLine("===Add Note===\n");
            Console.Write("Enter note title: ");
            title = Console.ReadLine();
            Console.Write("Enter note content: ");
            content = Console.ReadLine();

            encrypted_content = Login.encryption.Encrypt(content);

            try
            {
                using (StreamWriter writer = new StreamWriter("notes.txt", true))
                {
                    writer.WriteLine($"{title}|{encrypted_content}");
                }
            }
            catch (Exception) { Console.WriteLine("Error: file not found.\n"); }
        }
        public void getNote()
        {
            string title, crypted_content, decrypted_content;

            Console.WriteLine("\n  Secure Notes");
            Console.WriteLine("====Get Note====");
            Console.Write("Enter title: ");
            title = Console.ReadLine();

            try
            {
                using (StreamReader reader = new StreamReader("notes.txt"))
                {
                    string line;
                    bool found = false;

                    while ((line = reader.ReadLine()) != null)
                    {
                        string[] parts = line.Split('|');
                        if (parts.Length == 2 && parts[0] == title)
                        {
                            found = true;
                            crypted_content = parts[1];
                            decrypted_content = Login.encryption.Decrypt(crypted_content);
                            Console.WriteLine($"Note for {title}: {decrypted_content}\n");
                            break;
                        }
                    }
                    if (!found)
                    {
                        Console.WriteLine("No note found for that title.\n");
                    }
                }
            } catch(Exception) { Console.WriteLine("Error reading file.\n"); }
        }
        public void listNotes()
        {
            Console.WriteLine("\n  Secure Notes");
            Console.WriteLine("===List Notes===");

            try
            {
                using (StreamReader reader = new StreamReader("notes.txt"))
                {
                    string line;
                    int count = 0;
                    while ((line = reader.ReadLine()) != null)
                    {
                        string[] parts = line.Split('|');
                        if (parts.Length == 2)
                        {
                            count++;
                            Console.WriteLine($"[{count}] {parts[0]}");
                        }
                    }
                }
            } catch (Exception) { Console.WriteLine("Error: file not found.\n"); }
        }
        public void deleteNote()
        {
            string title, content;
            List<NoteEntry> Notes = new List<NoteEntry>();

            Console.WriteLine("\n  Secure Notes");
            Console.WriteLine("===Delete Note===");
            Console.Write("Enter title: ");
            title = Console.ReadLine();

            try
            {
                using (StreamReader reader = new StreamReader("notes.txt"))
                {
                    string line;
                    bool found = false;

                    while ((line = reader.ReadLine()) != null)
                    {
                        string[] parts = line.Split('|');
                        if (parts.Length == 2 && parts[0] != title)
                        {
                            Notes.Add(new NoteEntry
                            {
                                Title = parts[0],
                                Content = parts[1]
                            });
                        }
                        else if (parts.Length == 2 && parts[0] == title) found = true;
                    }
                    if (!found)
                    {
                        Console.WriteLine("No note found for that title.\n");
                    }

                }
            }
            catch (Exception) { Console.WriteLine("Error: file not found.\n"); }

            try
            {
                using (StreamWriter writer = new StreamWriter("notes.txt"))
                {
                    foreach (NoteEntry entry in Notes) { writer.WriteLine($"{entry.Title}|{entry.Content}"); }
                }
            }
            catch (Exception) { Console.WriteLine("Error: file not found.\n"); }
        }
    }
}
